src/core/matching_state.rs

```diff
@@ -1,78 +1,114 @@
+
+    /// Returns saved file paths for the current heuristic.
+    ///
+    /// This vector is inherited from parent directories and propagated while traversing filesystem.
+    /// It is used to check for matches in files that are not in the current directory
+    /// and/or store additional data for future calls.
+    pub fn inherited_files(&mut self) -> &mut Vec<PathBuf> {
+        self.inherited_files.entry(self.current_heuristic.type_id()).or_default()
+    }
+
+    /// Returns the path of the specified file in the current directory if it exists and is accesible.
+    ///
+    /// The result may be used to read the file contents.
     pub fn has_file(&self, name: &str) -> Option<PathBuf> {
-        self.contents.get(OsStr::new(name)).filter(|v| v.entry.file_type.is_file()).map(|v| v.entry.path())
+        self.contents.get(OsStr::new(name)).filter(|v| v.0.file_type.is_file()).map(|v| v.0.path())
     }
 
+    /// Returns the path of the specified directory in the current directory if it exists and is accesible.
+    ///
+    /// The result may be used to traverse the directory contents, but it is recommended to use
+    /// caching via [`inherited_files()`](Self::inherited_files()) instead (if possible).
     pub fn has_directory(&self, name: &str) -> Option<PathBuf> {
-        self.contents.get(OsStr::new(name)).filter(|v| v.entry.file_type.is_dir()).map(|v| v.entry.path())
+        self.contents.get(OsStr::new(name)).filter(|v| v.0.file_type.is_dir()).map(|v| v.0.path())
     }
 
+    /// Returns an iterator over all files in the current directory.
     pub fn get_all_directories(&self) -> impl Iterator<Item = PathBuf> + '_ {
-        self.contents.values().filter(move |v| v.entry.file_type.is_dir()).map(move |v| v.entry.path())
-    }
-
-    pub fn add_weight(&mut self, name: &str, weight: i32) {
-        if let Some(v) = self.contents.get_mut(OsStr::new(name)) {
-            v.weight += weight;
-            v.reasons.push(self.current_heuristic.unwrap().info());
-        }
+        self.contents.values().filter(move |v| v.0.file_type.is_dir()).map(move |v| v.0.path())
     }
 
-    pub fn add_weight_with_comment(&mut self, name: &str, weight: i32, comment: &str) {
-        if let Some(v) = self.contents.get_mut(OsStr::new(name)) {
-            v.weight += weight;
-            v.reasons.push(self.current_heuristic.unwrap().info().comment(comment));
+    /// Adds a match for the file or directory selected with the `name` parameter.
+    ///
+    /// The `comment` parameter is used to describe the match and is displayed to the user.
+    /// Additional match options may be changed by calling methods on the returned builder.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the specified file or directory does not exist in the current directory.
+    pub fn add_match(&mut self, name: &str, comment: &str) -> &mut MatchDataBuilder {
+        let new = MatchDataBuilder::new(self.current_heuristic.unwrap().info().with_comment(comment));
```

Comment:
that is weird builder patter. It would be nice to have `.build()` insted of wraping it in `MatchDataBuilder::new(...)`

Reply:
I am aware of that, but this way means less work for heuristic makers (so mainly future us I believe...). I use this weird pattern in order to spare the user some work with manually copying intended `LangData` from the heuristic's `info`, which would have been necessary if the builder was built by the user instead of by this function.

Reply:
https://github.com/Kacper0510/project_cleaner/pull/7#discussion_r1727473129